<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Assigning single-cells to known cell types with cellassign • cellassign</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Assigning single-cells to known cell types with cellassign">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cellassign</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction-to-cellassign.html">Assigning single-cells to known cell types with cellassign</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Assigning single-cells to known cell types with cellassign</h1>
                        <h4 class="author">Kieran R Campbell and Allen Zhang</h4>
            
            <h4 class="date">April 2018</h4>
      
      
      <div class="hidden name"><code>introduction-to-cellassign.Rmd</code></div>

    </div>

    
    
<p>CellAssign vignette skeleton. Missing descriptions for several components – to be added over the next while.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
<span class="co">#&gt; ── Attaching packages ──────────────────────────────────────────────────────────────────────────────────── tidyverse 1.2.1 ──</span>
<span class="co">#&gt; ✔ ggplot2 3.0.0     ✔ purrr   0.2.5</span>
<span class="co">#&gt; ✔ tibble  1.4.2     ✔ dplyr   0.7.6</span>
<span class="co">#&gt; ✔ tidyr   0.8.1     ✔ stringr 1.3.1</span>
<span class="co">#&gt; ✔ readr   1.1.1     ✔ forcats 0.3.0</span>
<span class="co">#&gt; ── Conflicts ─────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="co">#&gt; ✖ dplyr::filter() masks stats::filter()</span>
<span class="co">#&gt; ✖ dplyr::lag()    masks stats::lag()</span>
<span class="kw">library</span>(scater)
<span class="co">#&gt; Loading required package: SingleCellExperiment</span>
<span class="co">#&gt; Loading required package: SummarizedExperiment</span>
<span class="co">#&gt; Loading required package: GenomicRanges</span>
<span class="co">#&gt; Loading required package: stats4</span>
<span class="co">#&gt; Loading required package: BiocGenerics</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'BiocGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:parallel':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
<span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
<span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span>
<span class="co">#&gt; The following objects are masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     combine, intersect, setdiff, union</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     IQR, mad, sd, var, xtabs</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind,</span>
<span class="co">#&gt;     colMeans, colnames, colSums, dirname, do.call, duplicated,</span>
<span class="co">#&gt;     eval, evalq, Filter, Find, get, grep, grepl, intersect,</span>
<span class="co">#&gt;     is.unsorted, lapply, lengths, Map, mapply, match, mget, order,</span>
<span class="co">#&gt;     paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,</span>
<span class="co">#&gt;     Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,</span>
<span class="co">#&gt;     table, tapply, union, unique, unsplit, which, which.max,</span>
<span class="co">#&gt;     which.min</span>
<span class="co">#&gt; Loading required package: S4Vectors</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following objects are masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     first, rename</span>
<span class="co">#&gt; The following object is masked from 'package:tidyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid</span>
<span class="co">#&gt; Loading required package: IRanges</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'IRanges'</span>
<span class="co">#&gt; The following objects are masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     collapse, desc, slice</span>
<span class="co">#&gt; The following object is masked from 'package:purrr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     reduce</span>
<span class="co">#&gt; Loading required package: GenomeInfoDb</span>
<span class="co">#&gt; Loading required package: Biobase</span>
<span class="co">#&gt; Welcome to Bioconductor</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Vignettes contain introductory material; view with</span>
<span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="co">#&gt; Loading required package: DelayedArray</span>
<span class="co">#&gt; Loading required package: matrixStats</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'matrixStats'</span>
<span class="co">#&gt; The following objects are masked from 'package:Biobase':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyMissing, rowMedians</span>
<span class="co">#&gt; The following object is masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     count</span>
<span class="co">#&gt; Loading required package: BiocParallel</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'DelayedArray'</span>
<span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span>
<span class="co">#&gt; The following object is masked from 'package:purrr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     simplify</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     aperm, apply</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'scater'</span>
<span class="co">#&gt; The following object is masked from 'package:S4Vectors':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     rename</span>
<span class="co">#&gt; The following objects are masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     arrange, filter, mutate, rename</span>
<span class="co">#&gt; The following object is masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter</span>
<span class="kw">library</span>(splatter)
<span class="kw">library</span>(edgeR)
<span class="co">#&gt; Loading required package: limma</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'limma'</span>
<span class="co">#&gt; The following object is masked from 'package:scater':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     plotMDS</span>
<span class="co">#&gt; The following object is masked from 'package:BiocGenerics':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     plotMA</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'edgeR'</span>
<span class="co">#&gt; The following object is masked from 'package:SingleCellExperiment':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     cpm</span>
<span class="kw">library</span>(cellassign)</code></pre></div>
<div id="example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example data</h2>
<p>We’ll use example data generated by the <code>splatter</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(example_sce, <span class="dt">package =</span> <span class="st">"cellassign"</span>)</code></pre></div>
<p>This is a bit of a contrived example – the marker genes are VERY highly overexpressed in each group.</p>
<p>We can look at the true group assignments:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colData</span>(example_sce)<span class="op">$</span>Group <span class="op">%&gt;%</span><span class="st"> </span>table
<span class="co">#&gt; .</span>
<span class="co">#&gt; Group1 Group2 </span>
<span class="co">#&gt;    236    264</span></code></pre></div>
</div>
<div id="constructing-a-marker-gene-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#constructing-a-marker-gene-matrix" class="anchor"></a>Constructing a marker gene matrix</h2>
<p>The next step is to construct a binary marker gene matrix that expresses our <em>a priori</em> knowledge of cell types.</p>
<p>Since we’re using simulated data, we can pick genes that we made to be group-specific markers.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rowdat &lt;-<span class="st"> </span><span class="kw">rowData</span>(example_sce)

markers_group1 &lt;-<span class="st"> </span>rowdat <span class="op">%&gt;%</span><span class="st"> </span>as.data.frame <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(DEFacGroup1 <span class="op">&gt;</span><span class="st"> </span><span class="fl">1.3</span>)
markers_group2 &lt;-<span class="st"> </span>rowdat <span class="op">%&gt;%</span><span class="st"> </span>as.data.frame <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(DEFacGroup2 <span class="op">&gt;</span><span class="st"> </span><span class="fl">1.3</span>)</code></pre></div>
<p>Next, we’ll pick a few of these for each group, so as to not give CellAssign all the information.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">4242</span>)
genes1 &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">sample</span>(markers_group1<span class="op">$</span>Gene, <span class="dt">size =</span> <span class="dv">5</span>))
genes2 &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">sample</span>(markers_group2<span class="op">$</span>Gene, <span class="dt">size =</span> <span class="dv">5</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">marker_list &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="st">'Group1'</span>=genes1,
  <span class="st">'Group2'</span>=genes2
)</code></pre></div>
<p>We can conveniently convert this list into a binary matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rho &lt;-<span class="st"> </span><span class="kw"><a href="../reference/marker_list_to_mat.html">marker_list_to_mat</a></span>(marker_list, <span class="dt">include_other =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>Note that we have explicitly specified to not include an other class here (the default is to include one).</p>
</div>
<div id="running-cellassign" class="section level2">
<h2 class="hasAnchor">
<a href="#running-cellassign" class="anchor"></a>Running CellAssign</h2>
<div id="compute-size-factors" class="section level3">
<h3 class="hasAnchor">
<a href="#compute-size-factors" class="anchor"></a>Compute size factors</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">norm_factors &lt;-<span class="st"> </span>edgeR<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/edgeR/topics/calcNormFactors">calcNormFactors</a></span>(<span class="kw">as.matrix</span>(<span class="kw">counts</span>(example_sce)), <span class="dt">method =</span> <span class="st">"TMM"</span>)
lib_size_factors &lt;-<span class="st"> </span><span class="kw">colSums</span>(<span class="kw">as.matrix</span>(<span class="kw">counts</span>(example_sce)))

<span class="kw">sizeFactors</span>(example_sce) &lt;-<span class="st"> </span>norm_factors <span class="op">*</span><span class="st"> </span>lib_size_factors <span class="op">/</span><span class="st"> </span><span class="kw">mean</span>(norm_factors <span class="op">*</span><span class="st"> </span>lib_size_factors)

s &lt;-<span class="st"> </span><span class="kw">sizeFactors</span>(example_sce)</code></pre></div>
</div>
<div id="extract-expression-matrix" class="section level3">
<h3 class="hasAnchor">
<a href="#extract-expression-matrix" class="anchor"></a>Extract expression matrix</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Y &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">counts</span>(example_sce))[,<span class="kw">rownames</span>(rho)]</code></pre></div>
</div>
<div id="run-cellassign-em" class="section level3">
<h3 class="hasAnchor">
<a href="#run-cellassign-em" class="anchor"></a>Run CellAssign (EM)</h3>
<p>The following line runs the EM version of CellAssign.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">em_result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cellassign_em.html">cellassign_em</a></span>(<span class="dt">exprs_obj =</span> Y, <span class="dt">rho =</span> rho, <span class="dt">s =</span> s, <span class="dt">X =</span> <span class="ot">NULL</span>, <span class="dt">data_type =</span> <span class="st">"RNAseq"</span>, <span class="dt">n_batches =</span> <span class="dv">1</span>, <span class="dt">num_runs =</span> <span class="dv">1</span>, <span class="dt">learning_rate =</span> <span class="fl">1e-2</span>, <span class="dt">use_priors =</span> <span class="ot">TRUE</span>, <span class="dt">prior_type =</span> <span class="st">"shrinkage"</span>, <span class="dt">B =</span> <span class="dv">20</span>, <span class="dt">delta_variance_prior =</span> <span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt; 1360 L old: -41346.1149831637; L new: -13812.3255522503; Difference (%): 0.665934137756965</span>
<span class="co">#&gt; 220  L old: -13812.3255522503; L new: -13733.2130135937; Difference (%): 0.00572767694746749</span>
<span class="co">#&gt; 20   L old: -13733.2130135937; L new: -13732.1887213863; Difference (%): 7.45850374897777e-05</span>
<span class="co">#&gt; 20   L old: -13732.1887213863; L new: -13731.3764798901; Difference (%): 5.91487280460343e-05</span>
<span class="co">#&gt; 20   L old: -13731.3764798901; L new: -13730.7296108714; Difference (%): 4.71088255195157e-05</span>
<span class="co">#&gt; 20   L old: -13730.7296108714; L new: -13730.2113720548; Difference (%): 3.77429919076002e-05</span>
<span class="co">#&gt; 20   L old: -13730.2113720548; L new: -13729.7929559104; Difference (%): 3.04741225779837e-05</span>
<span class="co">#&gt; 20   L old: -13729.7929559104; L new: -13729.4518637258; Difference (%): 2.48432140014216e-05</span>
<span class="co">#&gt; 20   L old: -13729.4518637258; L new: -13729.1705904568; Difference (%): 2.04868535014388e-05</span>
<span class="co">#&gt; 20   L old: -13729.1705904568; L new: -13728.9355650361; Difference (%): 1.71186903953595e-05</span>
<span class="co">#&gt; 20   L old: -13728.9355650361; L new: -13728.7362973366; Difference (%): 1.45144318413023e-05</span>
<span class="co">#&gt; 20   L old: -13728.7362973366; L new: -13728.5646804444; Difference (%): 1.25005600251205e-05</span>
<span class="co">#&gt; 20   L old: -13728.5646804444; L new: -13728.4143662057; Difference (%): 1.0949013405643e-05</span>
<span class="co">#&gt; 20   L old: -13728.4143662057; L new: -13728.2800465636; Difference (%): 9.78406088788086e-06</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">em_result<span class="op">$</span>cell_type <span class="op">%&gt;%</span><span class="st"> </span>table
<span class="co">#&gt; .</span>
<span class="co">#&gt; Group1 Group2 </span>
<span class="co">#&gt;    236    264</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">example_sce &lt;-<span class="st"> </span>example_sce <span class="op">%&gt;%</span><span class="st"> </span>scater<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/scater/topics/mutate">mutate</a></span>(<span class="dt">EM_group=</span>em_result<span class="op">$</span>cell_type)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">with</span>(<span class="kw">colData</span>(example_sce), <span class="kw">table</span>(Group, EM_group))
<span class="co">#&gt;         EM_group</span>
<span class="co">#&gt; Group    Group1 Group2</span>
<span class="co">#&gt;   Group1    236      0</span>
<span class="co">#&gt;   Group2      0    264</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#example-data">Example data</a></li>
      <li><a href="#constructing-a-marker-gene-matrix">Constructing a marker gene matrix</a></li>
      <li><a href="#running-cellassign">Running CellAssign</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by First Last.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
